python/ray/train/v2/_internal/state/state_actor.py

```diff
@@ -1,47 +1,164 @@
         # This may be a memory issue for large runs.
-        self._runs: Dict[str, TrainRun] = {}
+        # TODO: consider cleaning up runs over time.
+        self._runs: Dict[str, TrainRun] = OrderedDict()
         # {run_id: {attempt_id: TrainRunAttempt}}
-        self._run_attempts: Dict[str, Dict[str, TrainRunAttempt]] = defaultdict(dict)
+        self._run_attempts: Dict[str, OrderedDict[str, TrainRunAttempt]] = defaultdict(
+            OrderedDict
+        )
         (
             self._export_logger,
             self._is_train_run_export_api_enabled,
             self._is_train_run_attempt_export_api_enabled,
         ) = self._init_export_logger()
 
+        # TODO: consider row level locking if loop takes too long.
+        self._runs_lock = threading.RLock()
+        self._run_attempts_lock = threading.RLock()
+
+        # Set env vars related to polling the controller.
+        if enable_state_actor_polling:
+            self._poll_interval_s = poll_interval_s
+            self._controllers_to_poll_per_iteration = controllers_to_poll_per_iteration
+            self._get_actor_timeout_s = get_actor_timeout_s
+            self._start_run_state_reconciliation_thread()
+
+    def _abort_live_runs_with_dead_controllers(
+        self, last_poll_run_id: Optional[str]
+    ) -> str:
+        aborted_run_ids = []
+        with self._runs_lock:
+            runs = list(self._runs.values())
+
+            # Start iterating from poll index.
+            poll_index = 0
+            if last_poll_run_id is not None:
+                for i, run in enumerate(runs):
+                    if run.id == last_poll_run_id:
+                        poll_index = (i + 1) % len(runs)
+                        break
+
+            # Abort runs.
+            num_sampled_runs = 0
+            i = poll_index
+            while (
+                i < poll_index + len(runs)
+                and num_sampled_runs < self._controllers_to_poll_per_iteration
+            ):
+                run = runs[i % len(runs)]
+                i += 1
+                last_poll_run_id = run.id
+                if run.status.is_terminal():
+                    continue
+                if not is_actor_alive(
+                    run.controller_actor_id, self._get_actor_timeout_s
+                ):
+                    update_train_run_aborted(run, False)
+                    self.create_or_update_train_run(run)
+                    aborted_run_ids.append(run.id)
+                num_sampled_runs += 1
+
+        # Abort run attempts.
+        with self._run_attempts_lock:
+            for run_id in aborted_run_ids:
+                latest_run_attempt = self._get_latest_run_attempt(run_id)
+                if latest_run_attempt and not latest_run_attempt.status.is_terminal():
+                    update_train_run_attempt_aborted(latest_run_attempt, False)
+                    self.create_or_update_train_run_attempt(latest_run_attempt)
```

Comment:
Any reason the attempt is processes in a separate loop rather than done together with the run?
```suggestion
                    update_train_run_aborted(run, False)
                    self.create_or_update_train_run(run)
                    latest_run_attempt = self._get_latest_run_attempt(run_id)
                        if latest_run_attempt and not latest_run_attempt.status.is_terminal():
                            update_train_run_attempt_aborted(latest_run_attempt, False)
                            self.create_or_update_train_run_attempt(latest_run_attempt)
                num_sampled_runs += 1
```

Reply:
I did it this way so we can use separate locks for runs and attempts - see the discussion here: https://github.com/ray-project/ray/pull/53818#discussion_r2162697915.